packages/datadog-instrumentations/src/http2/client.js

```diff
@@ -0,0 +1,67 @@
+'use strict'
+
+const shimmer = require('../../../datadog-shimmer')
+const { addHook, channel, AsyncResource } = require('../helpers/instrument')
+
+const startChannel = channel('apm:http2:client:request:start')
+const finishChannel = channel('apm:http2:client:request:finish')
+const errorChannel = channel('apm:http2:client:request:error')
+const responseChannel = channel('apm:http2:client:response')
+
+function createWrapEmit (requestResource, parentResource) {
+  return function wrapEmit (emit) {
+    return function (event, arg1) {
+      requestResource.runInAsyncScope(() => {
+        switch (event) {
+          case 'response':
+            responseChannel.publish(arg1)
+            break
+          case 'error':
+            errorChannel.publish(arg1)
+          case 'close': // eslint-disable-line no-fallthrough
+            finishChannel.publish()
+            break
+        }
+      })
+
+      return parentResource.runInAsyncScope(() => {
+        return emit.apply(this, arguments)
+      })
+    }
+  }
+}
+
+function createWrapRequest (authority, options) {
+  return function wrapRequest (request) {
+    return function (headers) {
+      const parentResource = new AsyncResource('bound-anonymous-fn')
+      const requestResource = new AsyncResource('bound-anonymous-fn')
+
+      return requestResource.runInAsyncScope(() => {
+        startChannel.publish({ headers, authority, options })
+
+        const req = request.apply(this, arguments)
+
+        shimmer.wrap(req, 'emit', createWrapEmit(requestResource, parentResource))
+
+        return req
+      })
+    }
+  }
+}
+
+function wrapConnect (connect) {
+  return function (authority, options) {
+    const session = connect.apply(this, arguments)
+
+    shimmer.wrap(session, 'request', createWrapRequest(authority, options))
```

Comment:
Do we _have_ to wrap on instances like this for request and emit? Do we not have access to these classes?

Reply:
It's not exported so no.

Reply:
Once we have an instance, we have its prototype, don't we? That is, we could shim the prototype lazily.

Reply:
This would make the code more complicated without a lot of benefit really since creating sessions is not something that would happen often. Given there are actually parameters from `connect` passed to the session, there would also need to be another way to store these based on the session. I don't think all the additional complexity is warranted here.